tufaceous/src/main.rs

```diff
@@ -0,0 +1,205 @@
+            )?;
+            update_versions(&mut editor, args.expiry)?;
+
+            let name = match &name {
+                Some(name) => name,
+                None => zone
+                    .file_name()
+                    .context("zone path is a directory")?
+                    .split('.')
+                    .next()
+                    .expect("str::split has at least 1 element"),
+            };
+            let filename = format!("{}-{}.tar.gz", name, version);
+
+            // if we already have an aritfact of this name/version/kind, replace it.
+            if let Some(artifact) =
+                artifacts.artifacts.iter_mut().find(|artifact| {
+                    artifact.name == name
+                        && artifact.version == version
+                        && artifact.kind == UpdateArtifactKind::Zone.into()
+                })
+            {
+                editor.remove_target(&artifact.target.as_str().try_into()?)?;
+                artifact.target = filename.clone();
+            } else {
+                // if we don't, make sure we're not overriding another target.
+                if existing_targets.iter().any(|target_name| {
+                    target_name.raw() == filename
+                        && target_name.resolved() == filename
+                }) {
+                    bail!(
+                        "a target named {:?} already exists in the repository",
+                        filename
+                    );
+                }
+                artifacts.artifacts.push(Artifact {
+                    name: name.to_owned(),
+                    version: version.clone(),
+                    kind: UpdateArtifactKind::Zone.into(),
+                    target: filename.clone(),
+                })
+            }
+
+            let targets_dir = repo_path.join("targets");
+
+            let mut file = TargetWriter::new(&targets_dir, filename)?;
+            std::io::copy(&mut File::open(&zone)?, &mut file)?;
+            file.finish(&mut editor)?;
+
+            let mut file = TargetWriter::new(&targets_dir, "artifacts.json")?;
+            serde_json::to_writer_pretty(&mut file, &artifacts)?;
+            file.finish(&mut editor)?;
+
+            let signed = editor.sign(&crate::key::boxed_keys(args.keys))?;
+            signed.write(repo_path.join("metadata"))?;
+            println!("added zone {}, version {}", name, version);
+            Ok(())
+        }
+    }
+}
+
+fn load_repo(repo_path: &Path) -> Result<Repository> {
+    let repo_path = fs::canonicalize(repo_path)?;
+    Ok(RepositoryLoader::new(
+        File::open(repo_path.join("metadata").join("1.root.json"))?,
+        Url::from_file_path(repo_path.join("metadata"))
+            .expect("the canonical path is not absolute?"),
+        Url::from_file_path(repo_path.join("targets"))
+            .expect("the canonical path is not absolute?"),
+    )
+    .expiration_enforcement(tough::ExpirationEnforcement::Unsafe)
+    .load()?)
```

Comment:
Not quite; the former calls `Into::into` on the error type, since we have a `Result<T, tough::error::Error>` and we need to return a `Result<T, anyhow::Error>`.

I think clippy does lint this if the error types are the same.

Reply:
Oh, that's cool. I didn't know about the error `into()`.